substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/InstrumentFeature.java

```diff
@@ -94,12 +97,34 @@ public void afterRegistration(AfterRegistrationAccess access) {
      * is absent. <br>
      * So this method looks for them in the same order.
      */
-    private void addPremainClass(String premainClass) {
+    private void addPremainClass(String javaagentOption) {
+        int separatorIndex = javaagentOption.indexOf("=");
+        String agent;
+        String premainClass = null;
+        String options = "";
+        // Get the agent file
+        if (separatorIndex == -1) {
+            agent = javaagentOption;
+        } else {
+            agent = javaagentOption.substring(0, separatorIndex);
+            options = javaagentOption.substring(separatorIndex + 1);
+        }
+        // Read MANIFEST in agent jar
+        try {
+            JarFile agentJarFile = new JarFile(agent);
+            premainClass = agentJarFile.getManifest().getMainAttributes().getValue("Premain-Class");
```

Comment:
From users' aspect, they need to manually isolate premain logic in JVM runtime and native image runtime anyway, a smoother and more convenient approach would be better.
Some of the agent premain logic could be shared between these two runtimes. Take OT agent for example, we carefully isolated class transformation and other staffs by runtime environment. The modifications are adding runtime environment checks for a few code sections, and most of the premain code is untouched. In this case, using a different method from `premain` will reduce the reusing of existing code. 

It is possible some simple agents don't have complicated framework to init in `premain` as OT agent. It is not necessary to re-execute the `premain` in native image. But the end users may not know such details. We should not try to educate them when to use which option to enable `premain` and when to omit it. It should leave to agent developer to decide how to implement the `premain`, and the end users just use it as what they do in JVM mode.

Reply:
@christianhaeubl the correct solution here would be that the Java agents don't do both transformation and initialization from the `premain` method, but have two methods instead. However, this would require a spec change which we can't do that easily. 

We are hoping here that most of the users won't need to re-write their agents as the `Instrumentation` we pass ignores all the method calls and says that no classes are transformable.

We are implementing this as an experimental feature to experiment, and if it works great. If most of the agents need to be rewritten, then I would go with a two-method solution that you propose.